Stop defining static data in Vec256

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37767 Stop defining static data in Vec256**

Fixes #37577

Needs tests, and maybe a lint.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21386704](https://our.internmc.facebook.com/intern/diff/D21386704)

ezyang committed:
Stop defining static data in Vec256

Fixes #37577

Needs tests, and maybe a lint.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Stop defining static data in Vec256"


Fixes #37577

Needs tests, and maybe a lint.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Stop defining static data in Vec256"


Fixes #37577

Needs tests, and maybe a lint.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

colesbury reviewed (approved):
LGTM. Is this sufficient to remove the AVX instructions from the initializer `_GLOBAL__sub_I_BinaryOpsKernel.cpp.AVX2.cpp` or are there other changes that will need to be made as well?

ezyang commented:
I need to check.

gchanan commented:
can you put comments at the top of the files to the effect of:

// DO NOT DEFINE STATIC DATA IN HERE

xuhdev commented on aten/src/ATen/cpu/vec256/vec256_bfloat16.h:
```diff
@@ -428,30 +427,28 @@ Vec256<BFloat16> inline operator^(const Vec256<BFloat16>& a, const Vec256<BFloat
   return _mm256_xor_si256(a, b);
 }
 
-const Vec256<BFloat16> Vec256<BFloat16>::ones(1.0f);
```
Can we create these constants statically inside each function that uses it? This should speed things up (no need to do the initialization step every time `ones` is needed) and it seems like they won't initialize until that function is executed for the first time: <https://stackoverflow.com/a/58804/1150462>

> An implementation is permitted to perform early initialization of other local objects with static storage duration under the same conditions that an implementation is permitted to statically initialize an object with static storage duration in namespace scope (basic.start.init). **Otherwise such an object is initialized the first time control passes through its declaration**; such an object is considered initialized upon the completion of its initialization.

colesbury replied:
No, please do not do this.

1) This is not an effective optimization. GCC/clang are already really good at lifting out these sorts of constants from inner loops.
2) It introduces atomic guards for the initialization. GCC/clang are not good at optimizing away the guards or lifting them out of loops.

xuhdev replied:
I now agree with Point 1 (after some experimentation). Mind explaining Point 2 a bit?

colesbury replied:
Here's an illustrative example. I've used `-Os` to make the assembly easier to read, but the same issue exists with `-O3`:

`and_ones_v1`: re-creates the "one" variable every time
`and_ones_v2`: uses function-local static data for the "one" variable

https://gcc.godbolt.org/z/gh4gps

Non-trivial static data needs to be initialized exactly once in a thread-safe manner. The compiler does this by adding a guard variable and a mutex. If the guard is set, the static data was already initialized. If it's not set, then it acquires the mutex, initializes the data and sets the guard variable. This is efficient for most C++ code, but can be slow for tight inner loops like in our kernels.

Current compilers aren't good at lifting the guard variable check outside the inner loop, so you end with at least an extra memory load and test+jmp in the inner loop. The generated call also seems to pessimize with other compiler analyses so the "one" variable is still reloaded every time through the loop.

In the hot-path through the loop `and_ones_v1` has one vector load (implicit in vandps) and one vector store. In the hot-path through the loop `and_ones_v2` has an extra vector load with `-O3` and two extra vector loads and an extra store in `-Os`.

xuhdev replied:
Thanks for the explanation, this is really helpful.

xuhdev commented on aten/src/ATen/cpu/vec256/vec256_bfloat16.h:
```diff
@@ -428,30 +427,28 @@ Vec256<BFloat16> inline operator^(const Vec256<BFloat16>& a, const Vec256<BFloat
-const Vec256<BFloat16> Vec256<BFloat16>::ones(1.0f);
-
 Vec256<BFloat16> Vec256<BFloat16>::eq(const Vec256<BFloat16>& other) const {
-  return (*this == other) & Vec256<BFloat16>::ones;
```
I meant adding a line like this here:

    static const Vec256<BFloat16> ones(1.0f);

ezyang replied:
This is still going to introduce an atomic guard for initialization

xuhdev reviewed (commented):
Awesome! Didn't realize static data here could cause running AVX code on non-AVX CPUs.

ezyang commented:
I verified that the "bad" symbol from the issue now compiles without AVX instructions:

```
00000000000e468d <_GLOBAL__sub_I_BinaryOpsKernel.cpp.AVX2.cpp>:
   e468d:>------55                   >--push   %rbp
   e468e:>------48 89 e5             >--mov    %rsp,%rbp
   e4691:>------be ff ff 00 00       >--mov    $0xffff,%esi
   e4696:>------bf 01 00 00 00       >--mov    $0x1,%edi
   e469b:>------e8 68 ff ff ff       >--callq  e4608 <_Z41__static_initialization_and_destruction_0ii>
   e46a0:>------5d                   >--pop    %rbp
   e46a1:>------c3                   >--retq···
```

I checked using `objdump -d ./build/caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp.AVX2.cpp.o`

ezyang committed:
Update on "Stop defining static data in Vec256"


Fixes #37577

Needs tests, and maybe a lint.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Stop defining static data in Vec256"


Fixes #37577

Needs tests, and maybe a lint.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21386704](https://our.internmc.facebook.com/intern/diff/D21386704)

[ghstack-poisoned]
